### ISSUE TITLE
Add feature for ordered parallel up based on dependencies

### ISSUE THREAD

**Issue by joejulian** [mitchellh/vagrant#5995]
It would be nice to be able to set dependencies for vms and have a parallel up bring up vms in groups to allow those dependencies to be met before starting the next group. 

For instance (and this is a poor example but...), webservers (web0 through web3) that require clustered storage (gluster0 through gluster3) and database (db0), database server that also requires clustered storage, and all of them requiring a salt master (salt0) as a development environment to emulate the hardware install.

Since they all have a dependency on `salt0` it ups and provisions first. The next iteration allows `gluster0` through `gluster3` to up and provision in parallel. Once those are done, `db0` ups and provisions. Finally `web0` through `web3` up and provision in parallel.

This would be defined by a vm config option, `require` that takes a list of host names (or maybe is a method that could be called repeatedly).

**Comment by sethvargo** [comment]
Hi @joejulian 

This sounds pretty similar to https://github.com/mitchellh/vagrant/issues/5596, do you agree? Maybe we could merge those two issues into a single feature request? Thoughts?

**Comment by joejulian** [comment]
IMHO, it is not. It appears that his need was starting N of xyz whereas this matches with the basic philosophy of, "Once you or someone else creates a single Vagrantfile, you just need to vagrant up and everything is installed and configured for you to work." where the entire system is upped in the sequence required.

The difference between this concept and that of the current implementation is that now we would have to disable parallel booting in order to satisfy dependencies so long as we add the VMs in the order required. My suggestion would allow vagrant to determine which VMs would need booted in what order and allow parallel booting when possible.

**Comment by sethvargo** [comment]
This is a duplicate of https://github.com/mitchellh/vagrant/issues/4026. Let's track that please. Thanks!
